### ISSUE TITLE
feat: dig into ionic animations / gestures

### ISSUE THREAD

**Issue by EinfachHans** [ionic-team/ionic-framework#23955]
### Prequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already include this feature request, without success.

### Describe the Feature Request

If a FR like this already exist, please close, but i didn't found one.

It would be great to be able to "dig" into Animations and Gestures that are used in the Ionic Framework.

### Describe the Use Case

It would allow us to show, hide or transform & animate stuff in the app while the user performs something. For example with the new sheet style modal we could fade content in or out if the modal is dragged up or dragged down.

Also this can be used for navigation animations.

### Describe Preferred Solution

I guess this is kinda tricky, as this can be either defined globally or per component. 

For example if i want to implement this for a Modal it must be possible for me to add this in the Component that the Modal uses like 

```
this.modal.animation.subscribe()
```

### Describe Alternatives

Currently i think something like this is possible if we copy/paste all the animations/gestures from ionic into our app and modify to call a handler, but this is not very good, as it doesn't update the animations if you update it internally 🤔

**Comment by liamdebeasi** [comment]
Can you please clarify what it means to "dig into an animation" in Ionic? It sounds like you are looking to get drag events emitted from the modal with the sheet gesture?

**Comment by EinfachHans** [comment]
Yes exactly. But not only the *Drag* Gesture and not only the Sheet Modal. If for example (when possible in the future) the sheet modal is changed via a function it should also executy my listener to perform my animations. 

Or if we talk about Page Transistions this should work on the first navigate, as well on the navigate back or swipe to go back.

**Comment by liamdebeasi** [comment]
Some of these animations/transitions are not meant to be extended/customized, so I would prefer not to add this feature to every animation unless we have use cases for each one.

The sheet modal use case makes sense, so I think we can add that one.

**Comment by EinfachHans** [comment]
Yeah maybe every animation was too much 😃 But i wouldn't limit this on sheet modals, we could start with modals in generall?

**Comment by liamdebeasi** [comment]
How would regular modals benefit from the drag event?

**Comment by EinfachHans** [comment]
I think of the Card style modal which is also swipable, so here it wouldn't be showing or hiding more or less content, it would be more having the posibility to do cool animations when draging the modal away

**Comment by liamdebeasi** [comment]
Thanks -- makes sense to have this work for card modal too.

**Comment by sean-perkins** [comment]
Added additional thoughts here: https://github.com/ionic-team/ionic-framework/issues/25127#issuecomment-1104375899

Summary: We should add a `dragEvents` property to `ion-modal`, to allow developers to subscribe to the drag events for both card and sheet modals. Sheet modals should likely include information related to the breakpoint as well.

**Comment by sean-perkins** [comment]
Found an example of where this feature could be leveraged in Instagram's app:

<video src="https://user-images.githubusercontent.com/13732623/173140650-5acfce46-72c9-4743-b1a9-059fd4a27193.mp4"></video>

They center the active location in the middle of the content behind the sheet modal, when the sheet modal begins to drag.

**Comment by JulienLecoq** [comment]
I also needs that to be able to hide dynamically some content when a modal is dragged 👍🏻

**Comment by sean-perkins** [comment]
Airbnb has UI that would require this feature, where the tab bar is moved off screen as the sheet modal is dragged.

https://github.com/ionic-team/ionic-framework/assets/13732623/9dd29c11-36a5-4381-904c-eec283ab606c

**Comment by greg-md** [comment]
In 2 years it was not implemented yet. :(

**Comment by aeharding** [comment]
Another use case: I would like to have a `dragStart` and `dragEnd` events on card modal to open and close the keyboard.

**Comment by Sovai** [comment]
2024 still waiting for dragStart/dragEnd events

**Comment by ijsje71** [comment]
I also need these drag events on the modals for another use case. Is this feature in the pipeline right now?

@thetaPC @ShaneK @brandyscarney

**Comment by derWebdesigner** [comment]
@thetaPC Having this feature would be amazing, dear Maria, because it would allow me to make my app a lot more interactive. I have a button on a map placed outside the modal (which is at 0.4 breakpoint) but I want to at least hide the button while the user is dragging the modal because having the button in the center of the screen, although the modal is dragged down, looks super weird. It would be even better if I could somehow attach the button to the modal and move it when the modal moves, but that doesn't seem to be possible either, because it's a different component and I can't place the button outside the component.

**Comment by thetaPC** [comment]
@derWebdesigner Once we have an update, we'll make sure to let you all know as soon as possible!

**Comment by EinfachHans** [comment]
@thetaPC is this closed with #30962 ?

**Comment by thetaPC** [comment]
@EinfachHans yes, thank you for catching this!

**Comment by thetaPC** [comment]
Completed through https://github.com/ionic-team/ionic-framework/pull/30962
